### PR TITLE
🪟 🎉  Connector Builder Landing Page

### LINKED ISSUE #21490: Connector Builder Landing Page
## Context

Implement the landing page for the connector builder, as shown in this design: https://www.figma.com/proto/F1ljvrJerLoVbNvtejBWid/LOW-CODE-CONNECTOR-BUILDER?page-id=3404%3A119487&node-id=3426%3A123290&viewport=-6573%2C-494%2C2&scaling=min-zoom&starting-point-node-id=3426%3A123290

**Note**: For now, the middle tile (Load an existing connector) and its corresponding page will not be added. This will be added once the "managing multiple connectors" project is complete.

## Acceptance Criteria

A landing page for the connector builder exists with two tiles:
- Import a YAML manifest, which opens the builder in the YAML view containing the imported yaml
- Start from scratch, which for now will just open a blank connector builder

Open question: should we always show this page when going to /connector-builder right now? Or should it only be shown if they don't have anything in local storage?
- Grooming notes: Go with the latter option here to get us incrementally there

### PR DESCRIPTION
**Pull request by lmossman**
## What
Resolves https://github.com/airbytehq/airbyte/issues/21490

Adds a landing page for the connector builder, which is shown if the user doesn't have any connector builder changes saved to browser local storage.

Screen Shot 2023-01-31 at 5 49 54 PM

To return to this page after making changes, click the Reset All button in the bottom-left corner of the Builder.

## How
As described in [the issue](https://github.com/airbytehq/airbyte/issues/21490), this PR only adds tiles for importing a YAML and starting from scratch. Start from scratch just opens a blank connector builder UI as normal.

The Import a YAML button only allows users to select a single `.yaml` file from their computer to upload. A few outcomes can occur here:
1. The uploaded file contains invalid YAML. In this case, we just show a Toast with the YAML validation error message and line number, which is auto-dismissed if they click the import button again.
2. The uploaded file contains valid YAML, but it can't be converted to Builder UI. In this case, we open the YAML in the YAML editor view of the connector builder.
3. The uploaded file contains valid YAML and can be converted to Builder UI. In this case, we open the Builder UI with the values filled in from the YAML.
    - In this case, the Connector Name is set to the titleCase version of the file name (minus the extension), e.g. `poke_api.yaml` would result in the connector name set to `Poke Api`

### CODE REVIEW

File: airbyte-webapp/src/components/connectorBuilder/StreamTestingPanel/ConfigMenu.tsx
Comment (lmossman) [inline airbyte-webapp/src/components/connectorBuilder/StreamTestingPanel/ConfigMenu.tsx:65]: Small bugfix - the page would crash if the `spec` did not contain `connection_specification`, and this change prevents that crash from occurring

File: airbyte-webapp/src/components/connectorBuilder/types.ts
Comment (lmossman) [inline airbyte-webapp/src/components/connectorBuilder/types.ts:102]: We want this eventually anyway, and this prevents us from showing a `Required` error on Reset All

File: airbyte-webapp/src/components/connectorBuilder/useManifestToBuilderForm.ts
Comment (lmossman) [inline airbyte-webapp/src/components/connectorBuilder/useManifestToBuilderForm.ts:64]: Fixes a bug where the lines below this one were actually updating the `DEFAULT_BUILDER_FORM_VALUES` constant's fields, causing issues like `Reset All` not behaving properly

File: airbyte-webapp/src/pages/ConnectorBuilderPage/LandingPage.tsx
Comment (lmossman) [inline airbyte-webapp/src/pages/ConnectorBuilderPage/LandingPage.tsx:131]: Open to suggestions here, but this seemed like the simplest and most standard way to handle file uploads without using a library. And I couldn't find any libraries that were well supported and lightweight

File: airbyte-webapp/src/pages/ConnectorBuilderPage/ConnectorBuilderPage.tsx
Comment (flash1293) [inline airbyte-webapp/src/pages/ConnectorBuilderPage/ConnectorBuilderPage.tsx:55]: As this is part of the regular `ConnectorBuilderPage`, it initializes the form and test state services and tries to do a first fetch which fails with a 400.

I wonder whether we should split out the landing page into its own page with its own path - it seems easier to understand in the code (no "page in a page") and we can better control what side effects are triggered and avoid state leaking from one connector to another in case the user switches (also I think we won't get around that for having multiple builder projects anyway). 

This is what I could imagine as interim state:
* `connector-builder/edit` shows the actual connector builder UI (redirects to `connector-builder` on reset)
* `connector-builder` shows the landing page (for now redirects automatically to `connector-builder/edit` if there is a non-default state in localstorage)

I think we can have a separate service for the landing page wiring up with local storage (and later on with the APIs to save and load connectors).


Doesn't absolutely need to happen on this PR as this is a meaningful increment, but it seems like a better step towards the targeted feature scope - your call how to go with this.

File: airbyte-webapp/src/pages/ConnectorBuilderPage/LandingPage.tsx
Comment (flash1293) [inline airbyte-webapp/src/pages/ConnectorBuilderPage/LandingPage.tsx:123]: nit: Why not moving all of this into the callback? Isolates a bit better the raw API access

File: airbyte-webapp/src/pages/ConnectorBuilderPage/LandingPage.tsx
Comment (flash1293) [inline airbyte-webapp/src/pages/ConnectorBuilderPage/LandingPage.tsx:121]: nit: the prop `hidden` should work as well here and is slightly more "correct" (not a big deal though)

File: airbyte-webapp/src/components/connectorBuilder/Builder/BuilderSidebar.tsx
Comment (flash1293) [inline airbyte-webapp/src/components/connectorBuilder/Builder/BuilderSidebar.tsx:74]: Doesn't need to be fixed on this PR, but it seems like this doesn't work reliably - after hidden "Reset all", then reloading the page I'm getting back to the previous state (there is probably another render-cycle of some kind putting it back). A separate page with a separate context for state management (see comment above) might help with this automatically.

File: airbyte-webapp/src/services/connectorBuilder/ConnectorBuilderStateService.tsx
Comment (flash1293) [inline airbyte-webapp/src/services/connectorBuilder/ConnectorBuilderStateService.tsx:98]: nit: `isEqual` isn't super cheap, would be nice to wrap it into an initializer function

**Review by flash1293 — COMMENTED** [review]
Looks mostly good besides some nits and one question about the code structure. Works like a charm!

File: airbyte-webapp/src/pages/ConnectorBuilderPage/ConnectorBuilderPage.tsx
Comment (lmossman) [reply airbyte-webapp/src/pages/ConnectorBuilderPage/ConnectorBuilderPage.tsx]: I like your suggestions here - I went ahead and added those changes now. I moved the local storage handling to its own ConnectorBuilderLocalStorage service/context, and moved the landing page into its own page / updated the routes and used redirects to go between.

I also found and fixed the issue with Reset All not always working properly - this was because it wasn't updating local storage properly, so I made sure it does that before navigating back to the landing page.

File: airbyte-webapp/src/pages/ConnectorBuilderPage/LandingPage.tsx
Comment (lmossman) [reply airbyte-webapp/src/pages/ConnectorBuilderPage/LandingPage.tsx]: I've moved this into the callback. The callback is now pretty long but I agree that keeping the logic together is probably better encapsulation

File: airbyte-webapp/src/components/connectorBuilder/Builder/BuilderSidebar.tsx
Comment (lmossman) [reply airbyte-webapp/src/components/connectorBuilder/Builder/BuilderSidebar.tsx]: See earlier comment, after the refactor and making this properly update local storage, this should now be fixed.

File: airbyte-webapp/src/services/connectorBuilder/ConnectorBuilderStateService.tsx
Comment (lmossman) [reply airbyte-webapp/src/services/connectorBuilder/ConnectorBuilderStateService.tsx:96]: Moved this into the landing page component since we now only need to do this once on page load. I put it in a useEffect which should only run a single time since I'm using refs for the local storage values

File: airbyte-webapp/src/pages/connectorBuilder/ConnectorBuilderLandingPage/ConnectorBuilderLandingPage.tsx
Comment (flash1293) [inline airbyte-webapp/src/pages/connectorBuilder/ConnectorBuilderLandingPage/ConnectorBuilderLandingPage.tsx:151]: We should always unregister this when navigating away from the page, otherwise the user could try to upload a yaml, give up and go with the "from scratch" option:
Screenshot 2023-02-02 at 15 35 27

Can be done via return value of an effect hook
```

useEffect(() => {
  return () => unregisterNotificationById(YAML_UPLOAD_ERROR_ID);
}, [unregisterNotificationById]);
```

File: airbyte-webapp/src/pages/connectorBuilder/ConnectorBuilderLandingPage/ConnectorBuilderLandingPage.tsx
Comment (flash1293) [inline airbyte-webapp/src/pages/connectorBuilder/ConnectorBuilderLandingPage/ConnectorBuilderLandingPage.tsx:104]: This used to work really well until https://github.com/airbytehq/airbyte/pull/21771 - now it will default to "Manifest" which is not super helpful. I think the next best thing is the url base, but no strong opinion on this.

My suggestion:
* If the file name is not "manifest", use the current logic
* If we can turn into form values, use the base url

File: airbyte-webapp/src/pages/connectorBuilder/ConnectorBuilderLandingPage/ConnectorBuilderLandingPage.tsx
Comment (flash1293) [inline airbyte-webapp/src/pages/connectorBuilder/ConnectorBuilderLandingPage/ConnectorBuilderLandingPage.tsx:99]: Probably out of scope for this PR, but I think it would be very nice to have telemetry for the landing page as well. Wdyt?

**Review by flash1293 — APPROVED** [review]
Looks great, just left to nits but LGTM otherwise. No need for another round of review

File: airbyte-webapp/src/pages/connectorBuilder/ConnectorBuilderLandingPage/ConnectorBuilderLandingPage.tsx
Comment (lmossman) [reply airbyte-webapp/src/pages/connectorBuilder/ConnectorBuilderLandingPage/ConnectorBuilderLandingPage.tsx:151]: Good catch! Fixed

File: airbyte-webapp/src/pages/connectorBuilder/ConnectorBuilderLandingPage/ConnectorBuilderLandingPage.tsx
Comment (lmossman) [reply airbyte-webapp/src/pages/connectorBuilder/ConnectorBuilderLandingPage/ConnectorBuilderLandingPage.tsx:100]: I will put up a separate PR for that 👍

**Pull request merged**
